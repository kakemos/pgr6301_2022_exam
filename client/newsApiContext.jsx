import React from "react";
import { fetchJSON } from "./lib/fetchJSON";
import { postJSON } from "./lib/postJSON";
import { deleteJSON } from "./lib/deleteJSON";

export const NewsApiContext = React.createContext({
  async listFilteredArticles(query) {
    return await fetchJSON("/api/news?" + new URLSearchParams(query));
  },
  async addNewArticle(article) {
    return await postJSON("/api/news/new", article);
  },
  async fetchLogin() {
    return await fetchJSON("/api/login");
  },
  async registerLogin(provider, login) {
    return await postJSON(`/api/login/${provider}`, login);
  },
  async endSession() {
    const res = await fetch("/api/login", { method: "DELETE" });
    if (!res.ok) {
      throw new Error(`Failed to post ${res.status}: ${res.statusText}`);
    }
  },
  async showArticle(title) {
    return await fetchJSON("/api/news?" + new URLSearchParams(title));
  },
  async deleteArticle(title) {
    return await deleteJSON("/api/news/article?" + new URLSearchParams(title));
  },
});
