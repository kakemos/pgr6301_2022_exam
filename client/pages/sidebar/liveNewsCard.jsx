import React from "react";

export function LiveNewsCard({ news: { title, intro } }) {
  return (
    <>
      <div className={"live-news-card"}>
        <hr />
        <h4 className={"live-news-title"}>
          <a href={"/article/" + title}>{title}</a>
        </h4>
        <p className={"live-news-intro"}>{intro}</p>
      </div>
    </>
  );
}
