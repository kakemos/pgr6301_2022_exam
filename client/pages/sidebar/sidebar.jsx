import React, { useContext, useEffect, useState } from "react";
import { NewsApiContext } from "../../newsApiContext";
import { useLoading } from "../../lib/useLoading";
import { LiveNewsCard } from "./liveNewsCard";

export function Sidebar() {
  const { listFilteredArticles } = useContext(NewsApiContext);
  const [newsFeed, setNewsFeed] = useState([]);

  useEffect(() => {
    const ws = new WebSocket(window.location.origin.replace(/^http/, "ws"));
    ws.onmessage = (message) => {
      const { title, intro } = JSON.parse(message.data);
      setNewsFeed((oldState) => [...oldState, { title, intro }]);
    };
  }, []);

  const { loading, error, data } = useLoading(
    async () => await listFilteredArticles({}),
    []
  );

  if (loading) {
    return <div>Loading...</div>;
  }
  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div id="error-text">{error.toString()}</div>
      </div>
    );
  }

  return (
    <>
      <h1>Live newsfeed</h1>

      {[...newsFeed].reverse().map((news) => (
        <LiveNewsCard key={news.title} news={news} />
      ))}

      {data.map((oldNews) => (
        <LiveNewsCard key={oldNews.title} news={oldNews} />
      ))}
    </>
  );
}
