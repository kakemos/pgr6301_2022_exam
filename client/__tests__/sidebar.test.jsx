import React from "react";
import { act } from "react-dom/test-utils";
import ReactDOM from "react-dom";
import { NewsApiContext } from "../newsApiContext";
import { BrowserRouter } from "react-router-dom";
import { Sidebar } from "../pages/sidebar/sidebar";

describe("sidebar component", () => {
  it("shows news articles", async () => {
    const news = [{ title: "article 1" }, { title: "article 2" }];
    const domElement = document.createElement("div");
    await act(async () => {
      ReactDOM.render(
        <NewsApiContext.Provider value={{ listFilteredArticles: () => news }}>
          <BrowserRouter>
            <Sidebar />
          </BrowserRouter>
        </NewsApiContext.Provider>,
        domElement
      );
    });
    expect(
      Array.from(domElement.querySelectorAll("h4")).map((e) => e.innerHTML)
    ).toEqual([
      '<a href="/article/article 1">article 1</a>',
      '<a href="/article/article 2">article 2</a>',
    ]);
    expect(domElement.innerHTML).toMatchSnapshot();
  });

  it("shows empty fields error message", () => {});
});
