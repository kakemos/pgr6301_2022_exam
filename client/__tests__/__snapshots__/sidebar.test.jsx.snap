// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sidebar component shows news articles 1`] = `"<h1>Live newsfeed</h1><div class=\\"live-news-card\\"><hr><h4 class=\\"live-news-title\\"><a href=\\"/article/article 1\\">article 1</a></h4><p class=\\"live-news-intro\\"></p></div><div class=\\"live-news-card\\"><hr><h4 class=\\"live-news-title\\"><a href=\\"/article/article 2\\">article 2</a></h4><p class=\\"live-news-intro\\"></p></div>"`;
